### PR TITLE
Fix Vertical Alignment Regression

### PR DESCRIPTION
**Pull request by justinmc**
This fixes a regression found while doing the roll where inputs with OutlineInputBorder and symmetrical padding don't center.  This was caused by https://github.com/flutter/flutter/pull/34355.

The solution here is to treat OutlineInputBorder differently (as it was before the referenced PR) so that it ignores padding when centering, but considers padding when aligning elsewhere.

### CODE REVIEW

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 1]
```diff
@@ -1055,13 +1057,32 @@ class _RenderDecoration extends RenderBox {
...
+    // The three main alignments for the baseline when an outline is present are
+    //
+    //  * top (-1.0): topmost point considering padding.
+    //  * center (0.0): the absolute center of the input ignoring padding but
+    //      accommodating the border and floating label.
+    //  * bottom (1.0): bottommost point considering padding.
+    //
+    // That means that if the padding is uneven, center is not the exact
+    // midpoint of top and bottom. To account for this, the points are
+    // interpolated between independently.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/input_decorator.dart]: the above center and below center alignments are interpolated independently

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 2]
```diff
@@ -1090,11 +1111,38 @@ class _RenderDecoration extends RenderBox {
...
       containerHeight: containerHeight,
       inputBaseline: inputBaseline,
+      outlineBaseline: outlineBaseline,
       subtextBaseline: subtextBaseline,
       subtextHeight: subtextHeight,
     );
   }
 
+  // Interpolate between three stops using textAlignVertical. This is used to
+  // calculate the outline baseline, which ignores padding for its middle but
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/input_decorator.dart]: ignores padding when the alignment is middle and interpolates between the centered text box's top and the top of the content padding when the alignment is less than zero [etc...]

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 3]
```diff
@@ -1090,11 +1111,38 @@ class _RenderDecoration extends RenderBox {
...
+  // uses padding for its beginning and end.
+  @visibleForTesting
+  double _interpolateThree(double begin, double middle, double end, TextAlignVertical textAlignVertical) {
+    if (textAlignVertical.y <= 0) {
+      // It's possible for begin, middle, and end to not be in order because of
+      // excessive padding. Those cases are handled by using middle.
+      if (begin >= middle) {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/input_decorator.dart]: NICE catch

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/input_decorator.dart (hunk 1 above)
Comment (justinmc) [reply packages/flutter/lib/src/material/input_decorator.dart]: Thanks, I was struggling to not say "we" or "I" in this sentence haha.

**Comment by alocher** [comment]
@justinmc I'm not really sure how, but is it possible that this change caused webview_flutter to no longer vertically center web content correctly? This seems to have started with 1.7.6 and is fine in 1.7.5.

Simple example with a webview going to a page with vertical centering. 1.7.5 shows correctly and 1.7.6 pushes everything to the top.

```
return Container(
        child: WebView(
      initialUrl: 'https://www.w3.org/Style/Examples/007/center-example.en.html'),
    ));
```

**Comment by justinmc** [comment]
Weird, I wouldn't think so...  @collinjackson is it possible there is any connection between InputDecorator alignment and webview_flutter?

**Pull request merged**